File: hotseat/sessions.py
"""Count Claude Code processes currently running on this machine.

This number is the whole reason the dashboard treats switching as dangerous:
changing the machine-wide default retargets every one of them mid-conversation.
"""

from __future__ import annotations

import os
import subprocess


def running_sessions() -> int:
    """Best-effort count of live `claude` processes owned by this user.

    Returns 0 rather than raising if the process list cannot be read: a wrong
    count must never stop the dashboard from rendering.
    """
    try:
        result = subprocess.run(["ps", "-eo", "pid=,args="],
                                capture_output=True, text=True, timeout=10, check=False)
    except (OSError, subprocess.SubprocessError):
        return 0
    if result.returncode != 0:
        return 0

    mine = os.getpid()
    count = 0
    for line in result.stdout.splitlines():
        parts = line.strip().split(None, 1)
        if len(parts) != 2:
            continue
        pid_text, args = parts
        try:
            pid = int(pid_text)
        except ValueError:
            continue
        if pid == mine:
            continue
        executable = args.split()[0] if args.split() else ""
        name = os.path.basename(executable)
        # Match the CLI itself, not this dashboard or an editor holding the word.
        if name == "claude" or name.startswith("claude-"):
            if "hotseat" in args:
                continue
            count += 1
    return count

File: internal/claude/sessions.go
package claude

// Count Claude Code processes currently running on this machine.
//
// This number is the whole reason the dashboard treats switching as dangerous:
// changing the machine-wide default retargets every one of them mid-conversation.

import (
	"context"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const psTimeout = 10 * time.Second

// RunningSessions is a best-effort count of live `claude` processes owned by
// this user.
//
// Returns 0 rather than failing if the process list cannot be read: a wrong
// count must never stop the dashboard from rendering.
func RunningSessions() int {
	return RunningSessionsWith(nil)
}

// RunningSessionsWith is RunningSessions with an injectable command runner
// (nil means (*exec.Cmd).Output).
func RunningSessionsWith(run RunOutput) int {
	if run == nil {
		run = func(cmd *exec.Cmd) ([]byte, error) { return cmd.Output() }
	}
	ctx, cancel := context.WithTimeout(context.Background(), psTimeout)
	defer cancel()
	out, err := run(exec.CommandContext(ctx, "ps", "-eo", "pid=,args="))
	if err != nil || ctx.Err() != nil {
		return 0
	}
	return CountSessions(string(out), os.Getpid())
}

// CountSessions parses `ps -eo pid=,args=` output, counting `claude` and
// `claude-*` executables other than the process `self` and anything that is
// part of hotseat itself.
func CountSessions(psOutput string, self int) int {
	count := 0
	for _, line := range strings.Split(psOutput, "\n") {
		fields := strings.Fields(line)
		if len(fields) < 2 {
			continue
		}
		pid, err := strconv.Atoi(fields[0])
		if err != nil {
			continue
		}
		if pid == self {
			continue
		}
		args := strings.Join(fields[1:], " ")
		name := filepath.Base(fields[1])
		// Match the CLI itself, not this dashboard or an editor holding the word.
		if name == "claude" || strings.HasPrefix(name, "claude-") {
			if strings.Contains(args, "hotseat") {
				continue
			}
			count++
		}
	}
	return count
}
